lib/instrumentation/modules/kafkajs.js

```diff
@@ -0,0 +1,451 @@
+        opts.tracestate = tracestate;
+      } else if (tracestate instanceof Buffer) {
+        opts.tracestate = tracestate.toString('utf-8');
+      }
+
+      const trans = ins.startTransaction(
+        `${NAME} RECEIVE from ${topic}`,
+        TYPE,
+        opts,
+      );
+
+      const messageCtx = { queue: { name: topic } };
+      if (
+        config.captureBody === 'all' ||
+        config.captureBody === 'transactions'
+      ) {
+        messageCtx.body = message.value.toString();
+      }
+
+      if (message.headers) {
+        messageCtx.headers = Object.keys(message.headers).reduce(
+          (acc, name) => {
+            const value = message.headers[name];
+            if (value instanceof Buffer) {
+              acc[name] = value.toString('utf-8');
+            } else {
+              acc[name] = value;
+            }
+            return acc;
+          },
+          {},
+        );
+      }
+
+      if (message.timestamp) {
+        messageCtx.age = {
+          ms: Date.now() - Number(message.timestamp),
+        };
+      }
+
+      trans.setMessageContext(messageCtx);
+
+      let result, err;
+      try {
+        result = await origEachMessage.apply(this, arguments);
+      } catch (ex) {
+        // Save the error for use in `finally` below, but re-throw it to
+        // not impact code flow.
+        err = ex;
+        throw ex;
+      } finally {
+        trans.setOutcome(
+          err ? constants.OUTCOME_FAILURE : constants.OUTCOME_SUCCESS,
+        );
+        trans.end();
+      }
+
+      return result;
+    };
+  }
+
+  /**
+   * Returns the instrumented version of `eachBatch` which
+   * - creates a transaction each time is called
+   * - if trace context present in messages inks them to the transaction
+   *
+   * @param {ConsumerRunConfig['eachBatch']} origEachBatch
+   * @returns {ConsumerRunConfig['eachBatch']}
+   */
+  function wrapEachBatch(origEachBatch) {
+    return async function ({ batch }) {
+      if (shouldIgnoreTopic(batch.topic, config)) {
+        return origEachBatch.apply(this, arguments);
+      }
+
+      const trans = ins.startTransaction(`${NAME} RECEIVE from batch`, TYPE);
+
+      // TODO: not sure if this should be here but we got batches for only one topic
+      // https://github.com/elastic/apm/blob/main/specs/agents/tracing-instrumentation-messaging.md#context-fields
+      const messageCtx = { queue: { name: batch.topic } };
+      trans.setMessageContext(messageCtx);
+
+      const serviceContext = {
+        framework: { name: 'Kafka' },
+      };
+      trans.setServiceContext(serviceContext);
+
+      // Extract span links from up to 1000 messages in this batch.
+      // https://github.com/elastic/apm/blob/main/specs/agents/tracing-instrumentation-messaging.md#receiving-trace-context
+      // A span link is created from a `traceparent` header in a message.
+      const messages = batch && batch.messages;
+
+      if (messages) {
+        const links = [];
```

Comment:
See my comment on the test case for this. Currently this is resulting in duplicate links (same trace_id and span_id) because it is receiving messages sent from the same outgoing span. Perhaps use a `Set` of seen traceparent values.

Reply:
changed!